src/main/java/org/javarosa/xpath/analysis/TopLevelContextTypesAnalyzer.java

```diff
@@ -0,0 +1,34 @@
+package org.javarosa.xpath.analysis;
+
+import org.javarosa.core.model.instance.TreeReference;
+
+/**
+ * An XPathAccumulatingAnalyzer that collects all of the TreeReference context types that are relevant
+ * to the evaluation of JUST the top level of the given expression, i.e. not including predicates
+ *
+ * @author Aliza Stone
+ */
+public class TopLevelContextTypesAnalyzer extends XPathAccumulatingAnalyzer<Integer> {
+
+    @Override
+    public void doAnalysis(TreeReference ref) throws AnalysisInvalidException {
+        addToResult(ref.getContextType());
+    }
+
+    @Override
+    XPathAnalyzer initSameTypeAnalyzer() {
+        return new TopLevelContextTypesAnalyzer();
+    }
+
+    @Override
+    public boolean shouldIncludePredicates() {
+        return false;
+    }
+
+    @Override
+    public boolean shortCircuit() {
+        // there are only 4 possibilities, so if we've gotten them all then no need to keep going
+        return size() == 4;
```

Comment:
For safety and future-proofing it might be worth including the set of potential answers as a list or enumeration in this class (or elsewhere) and using a count of that set. This seems like invisible/loose coupling to a concept that won't necessarily be clear when updated elsewhere.

Reply:
yea good call